CI is failing with:
```
2020-08-05T02:53:30.6429068Z Caused by: java.lang.IllegalArgumentException: Undeclared build item class io.quarkus.deployment.builditem.TransformedClassesBuildItem
2020-08-05T02:53:30.6429480Z 	at io.quarkus.builder.Messages.undeclaredItem(Messages.java:13)
2020-08-05T02:53:30.6430269Z 	at io.quarkus.builder.BuildResult.consume(BuildResult.java:48)
2020-08-05T02:53:30.6431699Z 	at io.quarkus.runner.bootstrap.StartupActionImpl.extractTransformers(StartupActionImpl.java:256)
2020-08-05T02:53:30.6432804Z 	at io.quarkus.runner.bootstrap.StartupActionImpl.<init>(StartupActionImpl.java:50)
2020-08-05T02:53:30.6434380Z 	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:133)
2020-08-05T02:53:30.6435314Z 	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:405)
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/builditem/TransformedClassesBuildItem.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/builditem/TransformedClassesBuildItem.java
@@ -36,12 +36,16 @@ public final class TransformedClassesBuildItem extends SimpleBuildItem {
 
     public static class TransformedClass {
 
+        private final String className;
         private final byte[] data;
         private final String fileName;
+        private final boolean eager;
 
-        public TransformedClass(byte[] data, String fileName) {
+        public TransformedClass(String className, byte[] data, String fileName, boolean eager) {
+            this.className = className;
             this.data = data;
             this.fileName = fileName;
+            this.eager = eager;
         }
 
         public byte[] getData() {
@@ -50,6 +54,14 @@ public final class TransformedClassesBuildItem extends SimpleBuildItem {
 
         public String getFileName() {
             return fileName;
+        }
+
+        public String getClassName() {
+            return className;
+        }
+
+        public boolean isEager() {
+            return eager;
         }
 
         @Override

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ClassTransformingBuildStep.java
@@ -45,6 +45,7 @@ public class ClassTransformingBuildStep {
                 bytecodeTransformerBuildItems.size());
         Set<String> noConstScanning = new HashSet<>();
         Map<String, Set<String>> constScanning = new HashMap<>();
+        Set<String> eager = new HashSet<>();
         for (BytecodeTransformerBuildItem i : bytecodeTransformerBuildItems) {
             bytecodeTransformers.computeIfAbsent(i.getClassToTransform(), (h) -> new ArrayList<>())
                     .add(i.getVisitorFunction());
@@ -53,6 +54,9 @@ public class ClassTransformingBuildStep {
             } else {
                 constScanning.computeIfAbsent(i.getClassToTransform(), (s) -> new HashSet<>())
                         .addAll(i.getRequireConstPoolEntry());
+            }
+            if (i.isEager()) {
+                eager.add(i.getClassToTransform());
             }
         }
         QuarkusClassLoader cl = (QuarkusClassLoader) Thread.currentThread().getContextClassLoader();
@@ -100,8 +104,8 @@ public class ClassTransformingBuildStep {
                                     visitor = i.apply(className, visitor);
                                 }
                                 cr.accept(visitor, 0);
-                                return new TransformedClassesBuildItem.TransformedClass(writer.toByteArray(),
-                                        classFileName);
+                                return new TransformedClassesBuildItem.TransformedClass(className, writer.toByteArray(),
+                                        classFileName, eager.contains(className));
                             } finally {
                                 Thread.currentThread().setContextClassLoader(old);
                             }

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java
@@ -32,7 +32,6 @@ import io.quarkus.builder.item.BuildItem;
 import io.quarkus.deployment.ExtensionLoader;
 import io.quarkus.deployment.QuarkusAugmentor;
 import io.quarkus.deployment.builditem.ApplicationClassNameBuildItem;
-import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.ConfigDescriptionBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedFileSystemResourceHandledBuildItem;
@@ -42,6 +41,7 @@ import io.quarkus.deployment.builditem.LiveReloadBuildItem;
 import io.quarkus.deployment.builditem.MainClassBuildItem;
 import io.quarkus.deployment.builditem.RawCommandLineArgumentsBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
+import io.quarkus.deployment.builditem.TransformedClassesBuildItem;
 import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
 import io.quarkus.deployment.pkg.builditem.JarBuildItem;
 import io.quarkus.deployment.pkg.builditem.NativeImageBuildItem;
@@ -56,8 +56,9 @@ public class AugmentActionImpl implements AugmentAction {
     private static final Logger log = Logger.getLogger(AugmentActionImpl.class);
 
     private static final Class[] NON_NORMAL_MODE_OUTPUTS = { GeneratedClassBuildItem.class,
-            GeneratedResourceBuildItem.class, BytecodeTransformerBuildItem.class, ApplicationClassNameBuildItem.class,
-            MainClassBuildItem.class, GeneratedFileSystemResourceHandledBuildItem.class };
+            GeneratedResourceBuildItem.class, ApplicationClassNameBuildItem.class,
+            MainClassBuildItem.class, GeneratedFileSystemResourceHandledBuildItem.class,
+            TransformedClassesBuildItem.class };
 
     private final QuarkusBootstrap quarkusBootstrap;
     private final CuratedApplication curatedApplication;
@@ -129,11 +130,9 @@ public class AugmentActionImpl implements AugmentAction {
             throw new IllegalStateException("Cannot launch a runtime application with NORMAL launch mode");
         }
         ClassLoader classLoader = curatedApplication.createDeploymentClassLoader();
-
         @SuppressWarnings("unchecked")
         BuildResult result = runAugment(true, Collections.emptySet(), classLoader, NON_NORMAL_MODE_OUTPUTS);
-
-        return new StartupActionImpl(curatedApplication, result, classLoader);
+        return new StartupActionImpl(curatedApplication, result);
     }
 
     @Override
@@ -146,7 +145,7 @@ public class AugmentActionImpl implements AugmentAction {
         @SuppressWarnings("unchecked")
         BuildResult result = runAugment(!hasStartedSuccessfully, changedResources, classLoader, NON_NORMAL_MODE_OUTPUTS);
 
-        return new StartupActionImpl(curatedApplication, result, classLoader);
+        return new StartupActionImpl(curatedApplication, result);
     }
 
     /**

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/StartupActionImpl.java
@@ -9,20 +9,15 @@ import java.lang.reflect.Method;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.StandardOpenOption;
-import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.function.BiFunction;
 import java.util.function.Consumer;
-import java.util.function.Predicate;
 
 import org.jboss.logging.Logger;
-import org.objectweb.asm.ClassVisitor;
 
 import io.quarkus.bootstrap.BootstrapDebug;
 import io.quarkus.bootstrap.app.CuratedApplication;
@@ -32,12 +27,11 @@ import io.quarkus.bootstrap.app.StartupAction;
 import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
 import io.quarkus.builder.BuildResult;
 import io.quarkus.deployment.builditem.ApplicationClassNameBuildItem;
-import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
 import io.quarkus.deployment.builditem.MainClassBuildItem;
+import io.quarkus.deployment.builditem.TransformedClassesBuildItem;
 import io.quarkus.deployment.configuration.RunTimeConfigurationGenerator;
-import io.quarkus.deployment.index.ConstPoolScanner;
 import io.quarkus.dev.appstate.ApplicationStateNotification;
 import io.quarkus.runtime.Quarkus;
 
@@ -49,35 +43,30 @@ public class StartupActionImpl implements StartupAction {
     private final BuildResult buildResult;
     private final QuarkusClassLoader runtimeClassLoader;
 
-    public StartupActionImpl(CuratedApplication curatedApplication, BuildResult buildResult,
-            ClassLoader deploymentClassLoader) {
+    public StartupActionImpl(CuratedApplication curatedApplication, BuildResult buildResult) {
         this.curatedApplication = curatedApplication;
         this.buildResult = buildResult;
         Set<String> eagerClasses = new HashSet<>();
-        Map<String, Predicate<byte[]>> transformerPredicates = new HashMap<>();
-        Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = extractTransformers(
-                eagerClasses, transformerPredicates);
+        Map<String, byte[]> transformedClasses = extractTransformers(eagerClasses);
         QuarkusClassLoader baseClassLoader = curatedApplication.getBaseRuntimeClassLoader();
         QuarkusClassLoader runtimeClassLoader;
 
         //so we have some differences between dev and test mode here.
         //test mode only has a single class loader, while dev uses a disposable runtime class loader
         //that is discarded between restarts
+        Map<String, byte[]> resources = new HashMap<>();
+        resources.putAll(extractGeneratedResources(true));
         if (curatedApplication.getQuarkusBootstrap().getMode() == QuarkusBootstrap.Mode.DEV) {
-            baseClassLoader.reset(extractGeneratedResources(false), bytecodeTransformers, transformerPredicates,
-                    deploymentClassLoader);
+            baseClassLoader.reset(extractGeneratedResources(false),
+                    transformedClasses);
             runtimeClassLoader = curatedApplication.createRuntimeClassLoader(baseClassLoader,
-                    bytecodeTransformers, transformerPredicates,
-                    deploymentClassLoader, extractGeneratedResources(true));
+                    resources, transformedClasses);
         } else {
-            Map<String, byte[]> resources = new HashMap<>();
             resources.putAll(extractGeneratedResources(false));
-            resources.putAll(extractGeneratedResources(true));
-            baseClassLoader.reset(resources, bytecodeTransformers, transformerPredicates, deploymentClassLoader);
+            baseClassLoader.reset(resources, transformedClasses);
             runtimeClassLoader = baseClassLoader;
         }
         this.runtimeClassLoader = runtimeClassLoader;
-        handleEagerClasses(runtimeClassLoader, eagerClasses);
     }
 
     private void handleEagerClasses(QuarkusClassLoader runtimeClassLoader, Set<String> eagerClasses) {
@@ -262,40 +251,18 @@ public class StartupActionImpl implements StartupAction {
         return runtimeClassLoader;
     }
 
-    private Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> extractTransformers(Set<String> eagerClasses,
-            Map<String, Predicate<byte[]>> transformerPredicates) {
-        Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = new HashMap<>();
-        Set<String> noConstScanning = new HashSet<>();
-        Map<String, Set<String>> constScanning = new HashMap<>();
-        List<BytecodeTransformerBuildItem> transformers = buildResult.consumeMulti(BytecodeTransformerBuildItem.class);
-        for (BytecodeTransformerBuildItem i : transformers) {
-            List<BiFunction<String, ClassVisitor, ClassVisitor>> list = bytecodeTransformers.get(i.getClassToTransform());
-            if (list == null) {
-                bytecodeTransformers.put(i.getClassToTransform(), list = new ArrayList<>());
-            }
-            list.add(i.getVisitorFunction());
-            if (i.isEager()) {
-                eagerClasses.add(i.getClassToTransform());
-            }
-            if (i.getRequireConstPoolEntry() == null || i.getRequireConstPoolEntry().isEmpty()) {
-                noConstScanning.add(i.getClassToTransform());
-            } else {
-                constScanning.computeIfAbsent(i.getClassToTransform(), (s) -> new HashSet<>())
-                        .addAll(i.getRequireConstPoolEntry());
-            }
-        }
-        for (String i : noConstScanning) {
-            constScanning.remove(i);
-        }
-        for (Map.Entry<String, Set<String>> entry : constScanning.entrySet()) {
-            transformerPredicates.put(entry.getKey(), new Predicate<byte[]>() {
-                @Override
-                public boolean test(byte[] bytes) {
-                    return ConstPoolScanner.constPoolEntryPresent(bytes, entry.getValue());
+    private Map<String, byte[]> extractTransformers(Set<String> eagerClasses) {
+        Map<String, byte[]> ret = new HashMap<>();
+        TransformedClassesBuildItem transformers = buildResult.consume(TransformedClassesBuildItem.class);
+        for (Set<TransformedClassesBuildItem.TransformedClass> i : transformers.getTransformedClassesByJar().values()) {
+            for (TransformedClassesBuildItem.TransformedClass clazz : i) {
+                ret.put(clazz.getFileName(), clazz.getData());
+                if (clazz.isEager()) {
+                    eagerClasses.add(clazz.getClassName());
                 }
-            });
+            }
         }
-        return bytecodeTransformers;
+        return ret;
     }
 
     private Map<String, byte[]> extractGeneratedResources(boolean applicationClasses) {

--- a/extensions/vertx-core/deployment/src/test/java/io/quarkus/vertx/core/deployment/VerticleWithClassNameDeploymentTest.java
+++ b/extensions/vertx-core/deployment/src/test/java/io/quarkus/vertx/core/deployment/VerticleWithClassNameDeploymentTest.java
@@ -33,7 +33,7 @@ public class VerticleWithClassNameDeploymentTest {
     public void testDeploymentOfVerticleUsingClassName() {
         String resp1 = RestAssured.get("http://localhost:8080").asString();
         String resp2 = RestAssured.get("http://localhost:8080").asString();
-        Assertions.assertTrue(resp1.startsWith("OK"));
+        Assertions.assertTrue(resp1.startsWith("OK"), resp1);
         Assertions.assertTrue(resp2.startsWith("OK"));
         Assertions.assertNotEquals(resp1, resp2);
     }

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java
@@ -19,11 +19,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.BiConsumer;
-import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Function;
-import java.util.function.Predicate;
-import org.objectweb.asm.ClassVisitor;
 
 /**
  * The result of the curate step that is done by QuarkusBootstrap.
@@ -269,19 +266,16 @@ public class CuratedApplication implements Serializable, Closeable {
     }
 
     public QuarkusClassLoader createRuntimeClassLoader(QuarkusClassLoader loader,
-            Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers,
-            Map<String, Predicate<byte[]>> transformerPredicates,
-            ClassLoader deploymentClassLoader, Map<String, byte[]> resources) {
+            Map<String, byte[]> resources, Map<String, byte[]> transformedClasses) {
         QuarkusClassLoader.Builder builder = QuarkusClassLoader.builder("Quarkus Runtime ClassLoader",
                 loader, false)
                 .setAggregateParentResources(true);
-        builder.setTransformerPredicates(transformerPredicates);
-        builder.setTransformerClassLoader(deploymentClassLoader);
+        builder.setTransformedClasses(transformedClasses);
 
+        builder.addElement(new MemoryClassPathElement(resources));
         for (Path root : quarkusBootstrap.getApplicationRoot()) {
             builder.addElement(ClassPathElement.fromPath(root));
         }
-        builder.addElement(new MemoryClassPathElement(resources));
 
         for (AdditionalDependency i : getQuarkusBootstrap().getAdditionalApplicationArchives()) {
             if (i.isHotReloadable()) {
@@ -290,7 +284,6 @@ public class CuratedApplication implements Serializable, Closeable {
                 }
             }
         }
-        builder.setBytecodeTransformers(bytecodeTransformers);
         return builder.build();
     }
 

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java
@@ -1,10 +1,7 @@
 package io.quarkus.bootstrap.classloading;
 
-import io.quarkus.bootstrap.BootstrapDebug;
 import java.io.ByteArrayInputStream;
 import java.io.Closeable;
-import java.io.File;
-import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.lang.reflect.InvocationTargetException;
@@ -23,14 +20,9 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
-import java.util.function.BiFunction;
-import java.util.function.Predicate;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
 import org.jboss.logging.Logger;
-import org.objectweb.asm.ClassReader;
-import org.objectweb.asm.ClassVisitor;
-import org.objectweb.asm.ClassWriter;
 
 /**
  * The ClassLoader used for non production Quarkus applications (i.e. dev and test mode). Production
@@ -75,10 +67,7 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
      *
      */
     private volatile MemoryClassPathElement resettableElement;
-
-    private volatile Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers;
-    private volatile Map<String, Predicate<byte[]>> transformerPredicates;
-    private volatile ClassLoader transformerClassLoader;
+    private volatile MemoryClassPathElement transformedClasses;
     private volatile ClassLoaderState state;
 
     static final ClassLoader PLATFORM_CLASS_LOADER;
@@ -105,15 +94,13 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         super(PLATFORM_CLASS_LOADER);
         this.name = builder.name;
         this.elements = builder.elements;
-        this.bytecodeTransformers = builder.bytecodeTransformers;
-        this.transformerPredicates = builder.transformerPredicates;
         this.bannedElements = builder.bannedElements;
         this.parentFirstElements = builder.parentFirstElements;
         this.lesserPriorityElements = builder.lesserPriorityElements;
         this.parent = builder.parent;
         this.parentFirst = builder.parentFirst;
         this.resettableElement = builder.resettableElement;
-        this.transformerClassLoader = builder.transformerClassLoader;
+        this.transformedClasses = new MemoryClassPathElement(builder.transformedClasses);
         this.aggregateParentResources = builder.aggregateParentResources;
     }
 
@@ -135,18 +122,13 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         return parentFirst || state.parentFirstResources.contains(name);
     }
 
-    public void reset(Map<String, byte[]> resources,
-            Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers,
-            Map<String, Predicate<byte[]>> transformerPredicates,
-            ClassLoader transformerClassLoader) {
+    public void reset(Map<String, byte[]> generatedResources, Map<String, byte[]> transformedClasses) {
         if (resettableElement == null) {
             throw new IllegalStateException("Classloader is not resettable");
         }
-        this.transformerClassLoader = transformerClassLoader;
         synchronized (this) {
-            this.transformerPredicates = transformerPredicates;
-            resettableElement.reset(resources);
-            this.bytecodeTransformers = bytecodeTransformers;
+            this.transformedClasses = new MemoryClassPathElement(transformedClasses);
+            resettableElement.reset(generatedResources);
             state = null;
         }
     }
@@ -234,11 +216,15 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
                                 throw new RuntimeException(
                                         "Resources cannot start with /, " + i + " is incorrect provided by " + element);
                             }
-                            List<ClassPathElement> list = elementMap.get(i);
-                            if (list == null) {
-                                elementMap.put(i, list = new ArrayList<>());
+                            if (transformedClasses.getResource(i) != null) {
+                                elementMap.put(i, Collections.singletonList(transformedClasses));
+                            } else {
+                                List<ClassPathElement> list = elementMap.get(i);
+                                if (list == null) {
+                                    elementMap.put(i, list = new ArrayList<>());
+                                }
+                                list.add(element);
                             }
-                            list.add(element);
                         }
                     }
                     Map<String, ClassPathElement[]> finalElements = new HashMap<>();
@@ -396,26 +382,6 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
                     ClassPathResource classPathElementResource = classPathElement.getResource(resourceName);
                     if (classPathElementResource != null) { //can happen if the class loader was closed
                         byte[] data = classPathElementResource.getData();
-                        List<BiFunction<String, ClassVisitor, ClassVisitor>> transformers = bytecodeTransformers.get(name);
-                        Map<String, Predicate<byte[]>> constPoolEntries = transformerPredicates;
-                        if (transformers != null) {
-                            data = handleTransform(name, data, transformers, constPoolEntries);
-
-                            if (BootstrapDebug.DEBUG_TRANSFORMED_CLASSES_DIR != null) {
-                                File debugPath = new File(BootstrapDebug.DEBUG_TRANSFORMED_CLASSES_DIR);
-                                if (!debugPath.exists()) {
-                                    debugPath.mkdir();
-                                }
-                                File classFile = new File(debugPath, name.replace(".", "/") + ".class");
-                                classFile.getParentFile().mkdirs();
-                                try (FileOutputStream classWriter = new FileOutputStream(classFile)) {
-                                    classWriter.write(data);
-                                } catch (Exception e) {
-                                    log.errorf(e, "Failed to write transformed class %s", name);
-                                }
-                                log.infof("Wrote transformed class to %s", classFile.getAbsolutePath());
-                            }
-                        }
                         definePackage(name, classPathElement);
                         return defineClass(name, data, 0, data.length,
                                 protectionDomains.computeIfAbsent(classPathElement, (ce) -> ce.getProtectionDomain(this)));
@@ -488,30 +454,6 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         return ret;
     }
 
-    private byte[] handleTransform(String name, byte[] bytes,
-            List<BiFunction<String, ClassVisitor, ClassVisitor>> transformers,
-            Map<String, Predicate<byte[]>> transformerPredicates) {
-        if (transformerPredicates.containsKey(name)) {
-            if (!transformerPredicates.get(name).test(bytes)) {
-                return bytes;
-            }
-        }
-        ClassReader cr = new ClassReader(bytes);
-        ClassWriter writer = new ClassWriter(cr, ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS) {
-
-            @Override
-            protected ClassLoader getClassLoader() {
-                return transformerClassLoader;
-            }
-        };
-        ClassVisitor visitor = writer;
-        for (BiFunction<String, ClassVisitor, ClassVisitor> i : transformers) {
-            visitor = i.apply(name, visitor);
-        }
-        cr.accept(visitor, 0);
-        return writer.toByteArray();
-    }
-
     @SuppressWarnings("unused")
     public Class<?> visibleDefineClass(String name, byte[] b, int off, int len) throws ClassFormatError {
         return super.defineClass(name, b, off, len);
@@ -543,11 +485,9 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         final List<ClassPathElement> bannedElements = new ArrayList<>();
         final List<ClassPathElement> parentFirstElements = new ArrayList<>();
         final List<ClassPathElement> lesserPriorityElements = new ArrayList<>();
-        Map<String, Predicate<byte[]>> transformerPredicates = Collections.emptyMap();
-        Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers = Collections.emptyMap();
         final boolean parentFirst;
         MemoryClassPathElement resettableElement;
-        private volatile ClassLoader transformerClassLoader;
+        private Map<String, byte[]> transformedClasses = Collections.emptyMap();
         boolean aggregateParentResources;
 
         public Builder(String name, ClassLoader parent, boolean parentFirst) {
@@ -643,29 +583,6 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
         }
 
         /**
-         * Sets any bytecode transformers that should be applied to this Class Loader
-         *
-         * @param bytecodeTransformers
-         */
-        public void setBytecodeTransformers(
-                Map<String, List<BiFunction<String, ClassVisitor, ClassVisitor>>> bytecodeTransformers) {
-            if (bytecodeTransformers == null) {
-                this.bytecodeTransformers = Collections.emptyMap();
-            } else {
-                this.bytecodeTransformers = bytecodeTransformers;
-            }
-        }
-
-        /**
-         * Sets const pool entries that need to be present for a class to be transformed. This optimisation
-         * allows for transformation to be skipped if the scanner detects that no transformation is required.
-         */
-        public Builder setTransformerPredicates(Map<String, Predicate<byte[]>> transformerPredicates) {
-            this.transformerPredicates = transformerPredicates;
-            return this;
-        }
-
-        /**
          * If this is true then a getResources call will always include the parent resources.
          *
          * If this is false then getResources will not return parent resources if local resources were found.
@@ -677,8 +594,8 @@ public class QuarkusClassLoader extends ClassLoader implements Closeable {
             return this;
         }
 
-        public Builder setTransformerClassLoader(ClassLoader transformerClassLoader) {
-            this.transformerClassLoader = transformerClassLoader;
+        public Builder setTransformedClasses(Map<String, byte[]> transformedClasses) {
+            this.transformedClasses = transformedClasses;
             return this;
         }
 

--- a/independent-projects/bootstrap/maven-resolver/pom.xml
+++ b/independent-projects/bootstrap/maven-resolver/pom.xml
@@ -23,10 +23,6 @@
             <artifactId>slf4j-jboss-logging</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.ow2.asm</groupId>
-            <artifactId>asm</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.apache.maven</groupId>
             <artifactId>maven-embedder</artifactId>
         </dependency>

--- a/independent-projects/bootstrap/pom.xml
+++ b/independent-projects/bootstrap/pom.xml
@@ -41,7 +41,6 @@
         <jakarta.inject-api.version>1.0</jakarta.inject-api.version>
         <commons-lang.version>3.9</commons-lang.version>
         <guava.version>27.0.1-jre</guava.version>
-        <asm.version>8.0.1</asm.version>
         <shrinkwrap-depchain.version>1.2.6</shrinkwrap-depchain.version>
         <jboss-logmanager-embedded.version>1.0.4</jboss-logmanager-embedded.version>
         <slf4j-jboss-logging.version>1.2.0.Final</slf4j-jboss-logging.version>
@@ -61,11 +60,6 @@
     </modules>
     <dependencyManagement>
         <dependencies>
-            <dependency>
-                <groupId>org.ow2.asm</groupId>
-                <artifactId>asm</artifactId>
-                <version>${asm.version}</version>
-            </dependency>
             <dependency>
                 <groupId>com.google.guava</groupId>
                 <artifactId>guava</artifactId>

